src/algorithms/diamondsquare.jl

```diff
@@ -0,0 +1,258 @@
+"""
+    DiamondSquare
+
+This type generates a neutral landscape using the diamond-sqaures
+algorithm, which produces fractals with variable spatialautocorrelation.
+
+https://en.wikipedia.org/wiki/Diamond-square_algorithm
+
+The algorithm is named diamond-square because it is an iterative procedure of
+"diamond" and "square" steps.
+
+
+
+The degree of spatial autocorrelation is controlled by a parameter `H`,
+which varies from 0.0 (low autocorrelation) to 1.0 (high autocorrelation) --- note this is non-inclusive and H = 0 and H = 1 will not behavive as expected.
+The result of the diamond-square algorithm is a fractal with dimension D = 2 + H.
+
+A similar algorithm, midpoint displacement *TODO link to mpd in docs*, almost
+identically, except that in DiamondSquare, the ed
+"""
+struct DiamondSquare <: NeutralLandscapeMaker
+    H::Float64
+    function DiamondSquare(H::T) where {T <: Real}
+        @assert 0 < H && H < 1
+        new(H)
+    end
+    DiamondSquare() = new(0.5)
+
+end
+
+"""
+    MPD()
+
+    Creates a midpoint-displacement algorithm object `MPD`. The degree of spatial autocorrelation is controlled by a parameter `H`,
+    which varies from 0.0 (low autocorrelation) to 1.0 (high autocorrelation) --- note this is non-inclusive and H = 0 and H = 1 will not behavive as expected.
+"""
+struct MPD <: NeutralLandscapeMaker
+    H::Float64
+    function MPD(H::T) where {T <: Real}
+        @assert 0 < H && H < 1
+        new(H)
+    end
+    MPD() = new(0.5)
+end
+
+
+"""
+    _landscape!(mat, alg::Union{DiamondSquare, MPD}; kw...)
+
+    Check if `mat` is the right size and already initialized.
+    If mat is not the correct size (DiamondSquare can only run on a lattice of size NxN where N = (2^n)+1 for integer n),
+    allocates the smallest lattice large enough to contain `mat` that can run DiamondSquare.
+    Will initialize `mat` to all zeros before running DiamondSquare if it is not initialized already.
+"""
+function _landscape!(mat, alg::Union{DiamondSquare, MPD}; kw...) where {IT <: Integer}
+
+    rightSize::Bool = _isPowerOfTwo(size(mat)[1]-1) && _isPowerOfTwo(size(mat)[2]-1)
+    latticeSize::Int = size(mat)[1]
+
+    dsMat = mat
```

Comment:
is this required?

Reply:
It makes a copy of mat if necessary and avoids it otherwise. Quite nice I think